test/src/state_machine/state.test.cpp

```diff
@@ -0,0 +1,781 @@
+    telemetry_data.service_propulsion_go   = static_cast<bool>(rand() > (RAND_MAX / 2));
+    telemetry_data.emergency_stop_command  = static_cast<bool>(rand() > (RAND_MAX / 2));
+    telemetry_data.nominal_braking_command = static_cast<bool>(rand() > (RAND_MAX / 2));
+  }
+
+  //---------------------------------------------------------------------------
+  // State Machine States
+  //---------------------------------------------------------------------------
+
+  static void randomiseStateMachine(StateMachine &stm_data)
+  {
+    stm_data.critical_failure = static_cast<bool>(rand() > (RAND_MAX / 2));
+  }
+
+  static void generateAllPermutations(ModuleStatus &module_status, EmergencyBrakes &embrakes_data,
+                                      Navigation &nav_data, Batteries &batteries_data,
+                                      Telemetry &telemetry_data, Sensors &sensors_data,
+                                      Motors &motors_data)
+  {
+    constexpr int num_statuses = 4;
+    constexpr int num_modules  = 6;
+    for (int i = 0; i <= pow(static_cast<double>(num_statuses), static_cast<double>(num_modules));
+         i++) {
+      embrakes_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+      nav_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+      batteries_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+      telemetry_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+      sensors_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+      motors_data.module_status = static_cast<ModuleStatus>(i % num_modules);
+      i /= num_statuses;
+    }
+  }
+};
+
+//---------------------------------------------------------------------------
+//--------------------------------- TESTS -----------------------------------
+//---------------------------------------------------------------------------
+
+//---------------------------------------------------------------------------
+// Idle Tests
+//---------------------------------------------------------------------------
+
+TEST_F(IdleTest, handlesEmergency)
+{
+  Data &data  = Data::getInstance();
+  Idle *state = Idle::getInstance();
+
+  EmergencyBrakes embrakes_data;
+  Navigation nav_data;
+  Batteries batteries_data;
+  Telemetry telemetry_data;
+  Sensors sensors_data;
+  Motors motors_data;
+
+  for (int i = 0; i < TEST_SIZE; i++) {
+    Randomiser::randomiseEmbrakes(embrakes_data);
+    Randomiser::randomiseNavigation(nav_data);
+    Randomiser::randomiseTelemetry(telemetry_data);
+    Randomiser::randomiseMotors(motors_data);
+
+    data.setEmergencyBrakesData(embrakes_data);
+    data.setNavigationData(nav_data);
+    data.setTelemetryData(telemetry_data);
+    data.setMotorData(motors_data);
```

Comment:
Here (and in other tests) we are only randomising 4 modules but we're actually using 6. Why is that?

What's up with batteries and sensors?